modules/javafx.controls/src/test/java/test/javafx/scene/control/ResizeHelperTest.java

```diff
@@ -0,0 +1,311 @@
+                    combineColumns(table, ix, ct, id++);
+                    break;
+                default:
+                    throw new Error("?" + cmd);
+                }
+            } else {
+                throw new Error("?" + x);
+            }
+        }
+
+        return table;
+    }
+
+    protected static void combineColumns(TableView<String> t, int ix, int count, int name) {
+        TableColumn<String,String> tc = new TableColumn<>();
+        tc.setText("N" + name);
+
+        for (int i = 0; i < count; i++) {
+            TableColumn<String, String> c = (TableColumn<String, String>)t.getColumns().remove(ix);
+            tc.getColumns().add(c);
+        }
+        t.getColumns().add(ix, tc);
+    }
+
+    /** verify that a custom constrained resize policy can indeed be implemented using public APIs */
+    @Test
+    public void testCanImplementCustomResizePolicy() {
+        double WIDTH = 15.0;
+
+        // constrained resize policy that simply sets all column widths to WIDTH
+        class UserPolicy
+            extends ConstrainedColumnResizeBase
+            implements Callback<TableView.ResizeFeatures, Boolean> {
+
+            @Override
+            public Boolean call(TableView.ResizeFeatures rf) {
+                List<? extends TableColumnBase<?, ?>> columns = rf.getTable().getVisibleLeafColumns();
+                int sz = columns.size();
+                // new public method getContentWidth() is visible
+                double w = rf.getContentWidth();
+                for (TableColumnBase<?, ?> c: columns) {
+                    // using added public method setColumnWidth()
+                    rf.setColumnWidth(c, WIDTH);
+                }
+                return false;
+            }
+        }
+
+        Object[] spec = {
+            Cmd.ROWS, 3,
+            Cmd.COL,
+            Cmd.COL,
+            Cmd.COL,
+            Cmd.COL
+        };
+        TableView<String> table = createTable(spec);
+
+        UserPolicy policy = new UserPolicy();
+        table.setColumnResizePolicy(policy);
+        table.setPrefWidth(10);
+
+        // verify the policy is in effect
+
+        stageLoader = new StageLoader(new BorderPane(table));
+        Toolkit.getToolkit().firePulse();
+
+        for (TableColumn<?, ?> c: table.getColumns()) {
+            assertEquals(WIDTH, c.getWidth());
+        }
+
+        // resize and check again
+        table.setPrefWidth(10_000);
+        Toolkit.getToolkit().firePulse();
+
+        for (TableColumn<?, ?> c: table.getColumns()) {
+            assertEquals(WIDTH, c.getWidth());
+        }
+    }
+
+    /**
+     * Exhausive behavioral test.
+     *
+     * Goes through all the policies, all valid combinations of constraints,
+     * and widths increasing to MAX_WIDTH and back,
+     * checkint that the initial resize does not violate (min,max) constraints.
+     */
+    //@Test // this test takes too much time!
+    public void testWidthChange() {
```

Comment:
Perhaps a combination of a number of typical tree/table configurations and a pseudo-random set could be added.  I would rather add it as a part of [JDK-8299753](https://bugs.openjdk.org/browse/JDK-8299753), since the new algorithm (simulated annealing) would have the constraint checks embedded in it on one hand, and on the other it would need a different set of tests, for example to verify that it converges.

A LONG_RUNNING_TEST is a good idea in general, I agree.  The problem with this particular test, and the reason why i did not pursue it, is that it does not finish in 40 minutes, which is way too long for a test.  I would like to think about it a bit longer, if I may.

Reply:
Yes, I do think this needs additional thought / discussion (and definitely not for this release).